wandb/sdk/launch/agent/agent.py

```diff
@@ -456,45 +456,42 @@ def loop(self) -> None:
                     raise KeyboardInterrupt
                 if self.num_running_jobs < self._max_jobs:
                     # only check for new jobs if we're not at max
-                    for queue in self._queues:
-                        job = self.pop_from_queue(queue)
-                        if job:
-                            try:
-                                file_saver = RunQueueItemFileSaver(
-                                    self._wandb_run, job["runQueueItemId"]
-                                )
-                                if _is_scheduler_job(job.get("runSpec")):
-                                    # If job is a scheduler, and we are already at the cap, ignore,
-                                    #    don't ack, and it will be pushed back onto the queue in 1 min
-                                    if (
-                                        self.num_running_schedulers
-                                        >= self._max_schedulers
-                                    ):
-                                        wandb.termwarn(
-                                            f"{LOG_PREFIX}Agent already running the maximum number "
-                                            f"of sweep schedulers: {self._max_schedulers}. To set "
-                                            "this value use `max_schedulers` key in the agent config"
-                                        )
-                                        continue
-                                self.run_job(job, queue, file_saver)
-                            except Exception as e:
-                                wandb.termerror(
-                                    f"{LOG_PREFIX}Error running job: {traceback.format_exc()}"
-                                )
-                                wandb._sentry.exception(e)
-
-                                # always the first phase, because we only enter phase 2 within the thread
-                                files = file_saver.save_contents(
-                                    contents=traceback.format_exc(),
-                                    fname="error.log",
-                                    file_sub_type="error",
-                                )
-                                self.fail_run_queue_item(
-                                    run_queue_item_id=job["runQueueItemId"],
-                                    message=str(e),
-                                    phase="agent",
-                                    files=files,
-                                )
+                    received_job = False
+                    for job, queue in self.get_jobs_and_queues():
```

Comment:
It looks like if an agent is polling multiple queues, this has the possibility to go over the max concurrency. For example if an agent is polling 3 queues, has a max jobs of 5, and is currently running 3 jobs, then `get_jobs_and_queues()` will grab one job from each queue and run it, bringing the agent to 6 jobs running.

Reply:
Oh nice catch, certainly an issue with our current implementation as well. I'll fix that

Reply:
Also consider queue "fairness" with your solution - if an agent is polling two queues and has a max jobs of 1, it shouldn't always pop from the first queue.

Reply:
Thats a known issue, but not in scope for this PR